test/FilenameUtilsTest.cpp

```diff
@@ -0,0 +1,510 @@
+#include "pch.h"
+#include <kf/FilenameUtils.h>
+
+using namespace kf;
+
+SCENARIO("FilenameUtils getPathNoEndSeparator")
+{
+    GIVEN("various file paths")
+    {
+        WHEN("getting path from empty path")
+        {
+            USimpleString emptyPath(L"");
+            auto result = FilenameUtils::getPathNoEndSeparator(emptyPath);
+
+            THEN("result is empty")
+            {
+                REQUIRE(result.isEmpty());
+            }
+        }
+
+        WHEN("getting path from path with several trailing slashes")
+        {
+            USimpleString pathWithTrailingSlashes(L"\\Device\\HarddiskVolume1\\Windows\\\\\\");
+            auto result = FilenameUtils::getPathNoEndSeparator(pathWithTrailingSlashes);
+
+            THEN("path is extracted without trailing separator")
+            {
+                REQUIRE(result.equals(USimpleString(L"\\Device\\HarddiskVolume1\\Windows\\\\")));
+            }
+        }
+
+        WHEN("getting path from path that consists only of slashes")
+        {
+            USimpleString slashesOnly(L"\\\\\\");
+            auto result = FilenameUtils::getPathNoEndSeparator(slashesOnly);
+
+            THEN("result is path without last slash")
+            {
+                REQUIRE(result.equals(USimpleString(L"\\\\")));
+            }
+        }
+
+        WHEN("getting path from path without any slashes")
+        {
+            USimpleString noSlashes(L"filename.txt");
+            auto result = FilenameUtils::getPathNoEndSeparator(noSlashes);
+
+            THEN("result is the same as input")
+            {
+                REQUIRE(result.equals(USimpleString(L"filename.txt")));
+            }
+        }
+
+        WHEN("getting path from normal NT path without trailing slash")
+        {
+            USimpleString normalPath(L"\\Device\\HarddiskVolume1\\Windows\\System32\\file.txt");
+            auto result = FilenameUtils::getPathNoEndSeparator(normalPath);
+
+            THEN("path is extracted correctly")
+            {
+                REQUIRE(result.equals(USimpleString(L"\\Device\\HarddiskVolume1\\Windows\\System32")));
```

Comment:
You should require that tne output is equal to the input because the function works like "get parent directory" instead just removing the trailing slash. Fix the function `getPathNoEndSeparator` according to new requirement.